### ISSUE TITLE
Inconsistency in Probability Values in TabularLIME

### ISSUE THREAD

**Issue by demirbilek95** [Azure/mmlspark#1049]
Hi,
I realize that there is an inconsistency among LIME probability and model probability. Is it about library or LIME is unable to approximate in this case (so local fidelity doesn't hold).

Before starting, I have a binary classification problem and I want to explain the results of the model by using the coefficients found by LIME.

**Steps that I followed:**
- I started with regular spark operations before model fitting. (e.g, stringindexing, vectorassembling)
- Later I fitted the gbtClassifier
- I used the fitted model to transform the data (to have probability and predictions from model)
- I have used TabularLIME to explain the result of the model, I fitted and transformed the same data. 
- At the end I have "probability_Model" and "probability" (from LIME)

Here is the screenshot of these two columns which is not even similar to each other

![Screenshot from 2021-05-12 11-14-54](https://user-images.githubusercontent.com/29048998/117950329-5c58b500-b313-11eb-9e86-b393bc824acd.png)

Here is the example code that I used:

`indexed = string_indexer.fit(df_cat).transform(df_cat)`
`output = assembler.transform(indexed)model = gbt.fit(output)` 

`pred = model.transform(output)`
`pred = pred.withColumnRenamed("rawPrediction","rawPrediction_Model").withColumnRenamed("probability","probability_Model").withColumnRenamed("prediction","prediction_Model")`

`lime = TabularLIME(model = model, inputCol="features", outputCol="weights", predictionCol = "prediction")`
`result = lime.fit(pred)`
`r = result.transform(pred)`

[AB#1167728](https://msdata.visualstudio.com/b9b2accc-2d1c-45b3-9d24-0eb5d78cc47f/_workitems/edit/1167728)

**Comment by demirbilek95** [comment]
Actually, I found similar problem in original library and over there there is a kernel_width parameter to overcome this issue but I haven't seen similar parameter in this library. Can you help about it?

**Comment by mhamilton723** [comment]
Hey @demirbilek95 if by kernel width you mean the size of the perturbative noise you can set this with:

  def setColumnSTDs(v: Array[Double])

on the TabularLIME model. Hope this helps and please let me know if this fixes your issue!

**Comment by demirbilek95** [comment]
> Hey @demirbilek95 if by kernel width you mean the size of the perturbative noise you can set this with:
> 
> def setColumnSTDs(v: Array[Double])
> 
> on the TabularLIMEModel (in your case the `result` object). Hope this helps and please let me know if this fixes your issue!

Thank you so much for the answer:
By **kernel_width**, I mean this:

> These local models operate in the proximity or neighborhood of the instance to be explained. They are fitted based on weights which indicate their proximity to the observation to be explained. The weights are typically determined using kernels that transform the proximity measure. [source](https://compstat-lmu.github.io/iml_methods_limitations/lime-neighbor.html)

There is also implementation in [lime](https://lime-ml.readthedocs.io/en/latest/lime.html#module-lime.lime_tabular). 
Hope we mean the same thing here.

I tried to set the columnSTD simply taking the std values for each column. But result didn't change a lot, it is better for the predictions 1, worse for predictions 0, maybe I need to set different values.

- How can we set good values for this parameter to have better approximations?
- What is the default values for this parameter?

![Screenshot from 2021-05-21 14-08-16](https://user-images.githubusercontent.com/29048998/119134924-0afca400-ba3e-11eb-9714-5571f0cc4d28.png)

Thank you so much for the help!

**Comment by mhamilton723** [comment]
@memoryz pointed out an issue in the lime implementation around the sampler and have recently queued a fix. Will reply here when its built with coordinates to try. Thanks for you patience!

**Comment by mhamilton723** [comment]
MMLSpark Build and Release Information
Maven Coordinates
com.microsoft.ml.spark:mmlspark_2.12:1.0.0-rc3-85-b4334027-SNAPSHOT

Maven Resolver
https://mmlspark.azureedge.net/maven

Documentation Pages:
Scala Documentation 
Python Documentation

**Comment by demirbilek95** [comment]
> @memoryz pointed out an issue in the lime implementation around the sampler and have recently queued a fix. Will reply here when its built with coordinates to try. Thanks for you patience!

Hey!
I tried with the release that you provided (both with setColumnSTDs and without it). I am not sure if it's the same issue. My problem still persists.

![Screenshot from 2021-05-25 16-42-52](https://user-images.githubusercontent.com/29048998/119517945-5b466f80-bd78-11eb-8fae-4e19cc699d34.png)

It could be a good starting point to understand how you handle (or if you have another approach, can you please provide detailed explanation about it?)  **kernel_width** because this parameter solved my issue when I decreased it in another library. I need similar behavior in spark context. For detailed explanation, also [this](https://compstat-lmu.github.io/iml_methods_limitations/lime-neighbor.html) can be checked.

> **kernel_width** – kernel width for the exponential kernel. [lime](https://lime-ml.readthedocs.io/en/latest/lime.html#module-lime.lime_tabular)

Thanks again for your help!

**Comment by memoryz** [comment]
> > @memoryz pointed out an issue in the lime implementation around the sampler and have recently queued a fix. Will reply here when its built with coordinates to try. Thanks for you patience!
> 
> Hey!
> I tried with the release that you provided (both with setColumnSTDs and without it). I am not sure if it's the same issue. My problem still persists.
> 
> ![Screenshot from 2021-05-25 16-42-52](https://user-images.githubusercontent.com/29048998/119517945-5b466f80-bd78-11eb-8fae-4e19cc699d34.png)
> 
> It could be a good starting point to understand how you handle (or if you have another approach to deal with it, can you please provide detailed explanation about it?) **kernel_width**, because this parameter solved my issue when I decreased the value of it in another library. I need similar behavior in spark context. For detailed explanation, also [this](https://compstat-lmu.github.io/iml_methods_limitations/lime-neighbor.html) can be checked.
> 
> > **kernel_width** – kernel width for the exponential kernel. [lime](https://lime-ml.readthedocs.io/en/latest/lime.html#module-lime.lime_tabular)
> 
> P.S. probability refers to lime results.
> 
> Thanks again for your help!

Hey @demirbilek95,

This issue is caused by a mixture of several things.

1. The current LIME explainer does not support categorical variables. It assumes all variables in the vector input are continuous. It perturbs a given instance by randomly sampling around the neighborhood of the instance by drawing from a normal distribution (The PR from last week (#1058) fixes a bug where the random sampling from the normal distribution wasn't implemented correctly). However, if your input variable is meant to be categorical, then the blackbox model (the gbt classifier in your case) will give unexpected results for the perturbed instances, making the interpretation unreliable.

2. The current LIME explainer does not support kernel width. The main reason is that the underlying regressor we used does not support sample weights.

3. The underlying regressor (aka. the local surrogate model) is a linear regression model, but it's fitting against the "prediction" column, and in your case, it's a binary output. The surrogate model may not work very well against binary response. It should fit better against the predicted probability.

4. I believe you misunderstood the "probability" column. The "probability" column is NOT the output of the LIME explainer, but rather is the blackbox's prediction of a certain perturbed instance. In other words, the "probability" column does not hold any real meaning, and should be ignored. The real output from the LIME explainer is the "weights" column (`outputCol="weights"`). The `weights` column contains the coefficients of the underlying local surrogate model. But due to the previous 3 problems, the `weights` output are quite likely still inconsistent.

We're currently actively working on the next version of the LIME explainer which will address problem 1, 2, and 3, as well as bringing additional features. As always, feedback like yours are greatly appreciated.

**Comment by memoryz** [comment]
> > > > @memoryz pointed out an issue in the lime implementation around the sampler and have recently queued a fix. Will reply here when its built with coordinates to try. Thanks for you patience!
> > > 
> > > 
> > > Hey!
> > > I tried with the release that you provided (both with setColumnSTDs and without it). I am not sure if it's the same issue. My problem still persists.
> > > ![Screenshot from 2021-05-25 16-42-52](https://user-images.githubusercontent.com/29048998/119517945-5b466f80-bd78-11eb-8fae-4e19cc699d34.png)
> > > It could be a good starting point to understand how you handle (or if you have another approach to deal with it, can you please provide detailed explanation about it?) **kernel_width**, because this parameter solved my issue when I decreased the value of it in another library. I need similar behavior in spark context. For detailed explanation, also [this](https://compstat-lmu.github.io/iml_methods_limitations/lime-neighbor.html) can be checked.
> > > > **kernel_width** – kernel width for the exponential kernel. [lime](https://lime-ml.readthedocs.io/en/latest/lime.html#module-lime.lime_tabular)
> > > 
> > > 
> > > P.S. probability refers to lime results.
> > > Thanks again for your help!
> 
> Thank you so much for this explanatory answer!
> 
> > Hey @demirbilek95,
> > This issue is caused by a mixture of several things.
> > 
> > 1. The current LIME explainer does not support categorical variables. It assumes all variables in the vector input are continuous. It perturbs a given instance by randomly sampling around the neighborhood of the instance by drawing from a normal distribution (The PR from last week (#1058) fixes a bug where the random sampling from the normal distribution wasn't implemented correctly). However, if your input variable is meant to be categorical, then the blackbox model (the gbt classifier in your case) will give unexpected results for the perturbed instances, making the interpretation unreliable.
> 
> * Actually, in this example, I didn't use categorical variable but I will have to use them. My question about this part is, how should we handle these categorical variables with current LIME implementation, is one-hot encoding a good idea? (actually this ends up with this exception `Caused by: ClassCastException: org.apache.spark.ml.linalg.SparseVector cannot be cast to org.apache.spark.ml.linalg.DenseVector` )
> 
> > 1. The current LIME explainer does not support kernel width. The main reason is that the underlying regressor we used does not support sample weights.
> 
> * Okay that explains many things so samples are equally effective on the explanations no matter how distant are they from the row that we want to explain.
> 
> > 1. The underlying regressor (aka. the local surrogate model) is a linear regression model, but it's fitting against the "prediction" column, and in your case, it's a binary output. The surrogate model may not work very well against binary response. It should fit better against the predicted probability.
> 
> * This totally make sense, but in current LIME implementation we only give the following parameters, **model** (fitted model)
>   **inputCol**,  **outputCol** , **predictionCol**. How can we fit against the predicted probability?
>   Because, input column is same as model input column, namely "features". Output columns is "weights", and I think it's just the name of the column. In my case prediction is given as "prediction" but isn't it just a name for the column again? Because when I had this column in my data (transformed from model) LIME gave me `IllegalArgumentException: requirement failed: Column prediction already exists.` so I thought  LIME uses **model parameter** to get the predictions (which I am not sure if it gets probability or binary prediction)
> 
> > 1. I believe you misunderstood the "probability" column. The "probability" column is NOT the output of the LIME explainer, but rather is the blackbox's prediction of a certain perturbed instance. In other words, the "probability" column does not hold any real meaning, and should be ignored. The real output from the LIME explainer is the "weights" column (`outputCol="weights"`). The `weights` column contains the coefficients of the underlying local surrogate model. But due to the previous 3 problems, the `weights` output are quite likely still inconsistent.
> 
> * Thanks for clarification, I know that "weights" the real output of LIME but I thought probability is the LIME prediction on that particular row, so I thought it should be more or less same as model prediction. As far as I understand you are saying that, this probability column is just a prediction of a perturbed instance (I didn't understand the certain part here maybe random) so it doesn't have to coincide with the model prediction?
> 
> > We're currently actively working on the next version of the LIME explainer which will address problem 1, 2, and 3, as well as bringing additional features. As always, feedback like yours are greatly appreciated.
> 
> I really appreciate this nice explanation, it is really helpful!
> 
> **Edit** : Two more questions, bringing the features to same scale is recommended? What about the correlation among the features, maybe regularization parameter helps to pick one of the correlated features?
> Thanks a lot!

- How should we handle these categorical variables with current LIME implementation, is one-hot encoding a good idea?

Unfortunately, one-hot encoding is not a good idea here. There are two reasons:
  1. The current LIME explainer assumes all inputs are continuous, and does not know that the one-hot encoded elements are binary only. So it would sample around the neighborhood of 0 or 1, and giving you fractions and negative values, which makes no sense to the underlying model.
  2. The current LIME explainer assumes feature independence. It does not know that the one-hot encoded vector (from the same categorical feature) can have only one "1". So the sample it draws wouldn't make sense to the underlying model, and will give you unreliable results.

At this moment, I'm afraid categorical variables are simply not supported. Our next version (in-progress) addresses this issue. 

- How can we fit against the predicted probability?

One thing you can try is to add a new transformer to your pipeline after the `gbt` stage to extract the class-1 probability, and use LIME to explain that extracted column. The easiest way to do this is to use a UDF transformer. See the `UnaryUDFTransformer` from this [SO answer](https://stackoverflow.com/a/35183614/3546203). The UDF you would need would look like this:
```scala
udf { (v: Vector, index: Int) => v(index) }
```
Again, in the next version, we will make this much easier to do.

- As far as I understand you are saying that, this probability column is just a prediction of a perturbed instance (I didn't understand the certain part here maybe random) so it doesn't have to coincide with the model prediction?

You are right. It's the probability of a random perturbed instance, and it doesn't have to coincide with the model prediction. The explainer's output is the `weights` column, and it should be interpreted as "how much will the model prediction (probability) increase/decrease if the corresponding feature is incremented/decremented by 1 unit, assuming the model behaves almost linearly in the local neighborhood".

- Bringing the features to same scale is recommended?

Very good question. In the current version, since kernel width (which determines sample weights based on the Euclidean distance between perturbed sample and original observation) is not supported, it doesn't really matter. In the next version, before we calculate the Euclidean distance, we normalize the distance by the feature's standard deviation, so feature scale shouldn't matter either (I haven't tested this theory yet).

- What about the correlation among the features, maybe regularization parameter helps to pick one of the correlated features?

The regularization parameter is set to the underlying LASSO regression's L1 penalty (the lambda parameter). It is used to remove features that are less correlated with the label. If you have two features that are 100% correlated, in theory the regularization parameter will not help you select one of them. But in practice, due to the randomness in sampling process, it is likely that regularization will still pick one feature at random. However that's not the purpose of the regularization parameter.

**Comment by demirbilek95** [comment]
Hey @memoryz!
Thank you so much for the informative answers!
Last two questions from my side, rest is crystal clear. 

> * How can we fit against the predicted probability?
> 
> One thing you can try is to add a new transformer to your pipeline after the `gbt` stage to extract the class-1 probability, and use LIME to explain that extracted column. The easiest way to do this is to use a UDF transformer. See the `UnaryUDFTransformer` from this [SO answer](https://stackoverflow.com/a/35183614/3546203). The UDF you would need would look like this:
> 
> ```scala
> udf { (v: Vector, index: Int) => v(index) }
> ```
> 
> Again, in the next version, we will make this much easier to do.
> 

Actually I tried this, but it gave me `llegalArgumentException: requirement failed: Column prediction already exists.`
Here is the code that I used with image of dataframe:

`indexed = string_indexer.fit(df_cat).transform(df_cat)`
`output = assembler.transform(indexed)model = gbt.fit(output) `
`pred = model.transform(output)`
`secondelement=udf(lambda v:float(v[1]),FloatType())`
`pred = pred.withColumn("probFraud",secondelement("probability"))`
`lime = TabularLIME(model = model, inputCol="features", outputCol="weights", predictionCol = "probFraud", regularization=1e-4, nSamples = 1500)`
`result = lime.fit(pred)`
`r = result.transform(pred)`

![Screenshot from 2021-05-31 10-24-39](https://user-images.githubusercontent.com/29048998/120164072-74c53c80-c1fa-11eb-93a1-2d3b4579ac6b.png)

Maybe I misunderstood what you described here, can you please point out the mistake that I am doing?

Last question is about the new version of LIME, can you please share the maven coordinates when new version is released? Also rough estimation of release date would be amazing.

Thank you so much for the help, it was really informative!

**Comment by memoryz** [comment]
Hey @demirbilek95,
> Last question is about the new version of LIME, can you please share the maven coordinates when new version is released? Also rough estimation of release date would be amazing.

Definitely. I'll update this thread with the PR when it's ready, so you can find out the maven coordinates once it's available. I cannot tell when the new version will be released or when the PR will be sent - my best estimates is 3-4 weeks. 

If you want to try it out first, head over to my fork at https://github.com/memoryz/mmlspark/tree/jasowang/lime. The new LIME implementation is pretty much in shape. Checkout the unit test (com/microsoft/ml/spark/explainers/LIMESuite.scala) to get started. However: 1. the API is still unstable as we add in more features. 2. no pyspark support yet, only the JVM based version is available. 3. You'd have to build the code from sbt.

**Comment by memoryz** [comment]
The PR was merged into the master branch.

Maven Coordinates
`com.microsoft.ml.spark:mmlspark_2.12:1.0.0-rc3-102-c84ab470-SNAPSHOT`

Maven Resolver
`https://mmlspark.azureedge.net/maven`

Documents and notebooks will come in next PR.
